src/contracts/SavingCircles.sol

```diff
@@ -0,0 +1,276 @@
+
+import {OwnableUpgradeable} from '@openzeppelin-upgradeable/access/OwnableUpgradeable.sol';
+import {IERC20} from '@openzeppelin/token/ERC20/IERC20.sol';
+import {ReentrancyGuard} from '@openzeppelin/utils/ReentrancyGuard.sol';
+
+import {ISavingCircles} from '../interfaces/ISavingCircles.sol';
+
+/**
+ * @title Saving Circles
+ * @notice Simple implementation of a rotating savings and credit association (ROSCA) for ERC20 tokens
+ * @author Breadchain Collective
+ * @author @RonTuretzky
+ * @author @bagelface
+ */
+contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+  uint256 public constant MINIMUM_MEMBERS = 2;
+
+  mapping(bytes32 id => Circle circle) public circles;
+  mapping(address token => bool status) public allowedTokens;
+  mapping(bytes32 id => mapping(address token => uint256 balance)) public balances;
+  mapping(bytes32 id => mapping(address member => bool status)) public isMember;
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(address _owner) external override initializer {
+    __Ownable_init_unchained(_owner);
+  }
+
+  /**
+   * @notice Create a new saving circle
+   * @param _circle A new saving circle
+   */
+  function create(Circle memory _circle) external override {
+    bytes32 _id = keccak256(abi.encodePacked(_circle.name));
+
+    if (circles[_id].owner != address(0)) revert AlreadyExists();
+    if (
+      !allowedTokens[_circle.token] || _circle.depositInterval == 0 || _circle.depositAmount == 0
+        || _circle.maxDeposits == 0 || _circle.circleStart == 0 || _circle.currentIndex != 0
+        || _circle.owner == address(0) || _circle.members.length < MINIMUM_MEMBERS
+    ) revert InvalidCircle();
+
+    for (uint256 i = 0; i < _circle.members.length; i++) {
+      address _member = _circle.members[i];
+      if (_member == address(0)) revert InvalidCircle();
+      isMember[_id][_member] = true;
+    }
+    circles[_id] = _circle;
+
+    emit CircleCreated(
+      _id, _circle.name, _circle.members, _circle.token, _circle.depositAmount, _circle.depositInterval
+    );
+  }
+
+  /**
+   * @notice Make a deposit into a specified circle
+   * @param _id Identifier of the circle
+   * @param _value Amount of the token to deposit
+   */
+  function deposit(bytes32 _id, uint256 _value) external override nonReentrant {
+    _deposit(_id, _value, msg.sender);
+  }
+
+  /**
+   * @notice Make a deposit on behalf of another member
+   * @param _id Identifier of the circle
+   * @param _value Amount of the token to deposit
+   * @param _member Address to make a deposit for
+   */
+  function depositFor(bytes32 _id, uint256 _value, address _member) external override nonReentrant {
+    _deposit(_id, _value, _member);
+  }
+
+  /**
+   * @notice Make a withdrawal from a specified circle
+   * @param _id Identifier of the circle
+   */
+  function withdraw(bytes32 _id) external override nonReentrant {
```

Comment:
As discussed, we might want to make a `withdrawFor` function in order to allow withdrawing for another member in the event they are unable (like the `depositFor` functionality)

Reply:
added